packages/btcstaking/src/scripts_utils.rs

```diff
@@ -1,106 +1,96 @@
-    let key_bytes = hex::decode(UNSPENDABLE_KEY).unwrap();
+    builder = builder.push_int(threshold as i64);
+    if with_verify {
+        builder = builder.push_opcode(OP_NUMEQUALVERIFY);
+    } else {
+        builder = builder.push_opcode(OP_NUMEQUAL);
+    }
 
-    let (pk_x, _) = PublicKey::from_slice(&key_bytes)
-        .unwrap()
-        .x_only_public_key();
-    pk_x
+    Ok(builder.into_script())
 }
 
-// sort_keys sorts public keys in lexicographical order
-pub fn sort_keys(keys: &mut [VerifyingKey]) {
+/// sort_keys sorts public keys in lexicographical order
+fn sort_keys(keys: &mut [VerifyingKey]) {
     keys.sort_by(|a, b| {
         let a_serialized = a.to_bytes();
         let b_serialized = b.to_bytes();
         a_serialized.cmp(&b_serialized)
     });
 }
 
-/// prepare_keys_for_multisig_script prepares keys for multisig, ensuring there are no duplicates
-pub fn prepare_keys_for_multisig_script(keys: &[VerifyingKey]) -> Result<Vec<VerifyingKey>> {
+/// prepare_keys_for_multisig_script prepares keys to be used in multisig script
+/// Validates whether there are at least 2 keys
+/// and returns copy of the slice of keys sorted lexicographically.
+///
+/// Note: It is up to the caller to ensure that the keys are unique
+fn prepare_keys_for_multisig_script(keys: &[VerifyingKey]) -> Result<Vec<VerifyingKey>> {
     if keys.len() < 2 {
         return Err(Error::InsufficientMultisigKeys {});
     }
 
     let mut sorted_keys = keys.to_vec();
     sort_keys(&mut sorted_keys);
 
-    // Check for duplicates
-    for window in sorted_keys.windows(2) {
-        if window[0] == window[1] {
-            return Err(Error::DuplicateKeys {});
-        }
-    }
-
     Ok(sorted_keys)
 }
 
-/// assemble_multisig_script assembles a multisig script
-fn assemble_multisig_script(
-    pubkeys: &[VerifyingKey],
-    quorum: usize,
+/// build_multisig_script creates multisig script with given keys and signer threshold to
+/// successfully execute script
+/// it validates whether threshold is not greater than number of keys
+/// If there is only one key provided it will return single key sig script
+/// Note: It is up to the caller to ensure that the keys are unique
+pub(crate) fn build_multisig_script(
+    keys: &[VerifyingKey],
+    threshold: usize,
     with_verify: bool,
 ) -> Result<ScriptBuf> {
-    if quorum > pubkeys.len() {
-        return Err(Error::QuorumExceedsKeyCount {});
+    if keys.is_empty() {
+        return Err(Error::NoKeysProvided {});
     }
 
-    let mut builder = Builder::new();
-    for (i, key) in pubkeys.iter().enumerate() {
-        let pk_bytes: [u8; 32] = key.to_bytes().into();
-        builder = builder.push_slice(pk_bytes);
-        if i == 0 {
-            builder = builder.push_opcode(OP_CHECKSIG);
-        } else {
-            builder = builder.push_opcode(OP_CHECKSIGADD);
-        }
+    if threshold > keys.len() {
+        return Err(Error::ThresholdExceedsKeyCount {});
```

Comment:
Nice to enrich the variant.

```suggestion
        return Err(Error::ThresholdExceedsKeyCount { threshold, keys_count: keys.len() });
```